Scripts in these folders require summary statistics from GWAS.

Scripts in *reported_refined_novel_variants/*: Scripts to classify lists of variants as reported (in previously published GWASs of 44 obesity-associated traits in the GWAS Catalog, accessed on 02 Nov 2021), in linkage disequlibrium with reported SNPs, or novel SNPs, based on conditional analysis criteria established by [Benonisdottir et al. 2016](https://www.nature.com/articles/ncomms13490).

- **classify_reported_novel_variants.R** - Read in list of GWS or fine-mapped SNPs for classification; write reported SNPs by cross-checking against GWAS Catalog obesity-associated SNPs, and create list of unreported SNPs for conditional analysis - prepare file by getting all reported variants within 500kb of each unreported SNP. Submit **gcta_cojo_conditional_analysis.sh** from within this file. Format for GCTA-COJO analysis and batch submit across the different strata with **submit_classify_reported_novel_variants.sh**.
- **gcta_cojo_conditional_analysis.sh** - For a given list of SNPs (containing a SNP of interest and all obesity-associated SNPs within 500kb), create LD matrix based on individuals of White British ancestry to perform conditional analysis with GCTA-COJO. All collinear SNPs are added to the "reported" SNP list; if the SNP of interest remains as an independent association, it may be refined or novel -- calculate conditional effect of all independent SNPs in the region.
- **characterise_refined_novel_variants.R** - From the conditional effects calculated above, classify SNPs as reported, refined, or novel based on the "significantly stronger effect" and "conditionally independent" criteria using t-tests. Access "buddy" SNPs for each refined and novel SNP, i.e. most highly correlated or nearest obesity-associated variant.
